README.md
# A Perl client library for Streamtools


## Synopsis

``` perl
  use Streamtools;

  # Connect to the Streamtools server at http://localhost:7070
  my $st = Streamtools->new(server => 'localhost', port => 7070);
```


## Description

This module provides a Perl client library for the REST API provided by
Streamtools.

From the developer blurb: "Streamtools is a graphical toolkit for dealing
with streams of data. Streamtools makes it easy to explore, analyse,
modify and learn from streams of data."

To find out more about Streamtools, have a look at
[their website](http://nytlabs.github.io/streamtools/).



CHANGES
Revision history for Perl module Streamtools:

{{$NEXT}}

t/lib/Streamtools/API/NullRequest.pm
use strict;
use warnings;
package Streamtools::API::NullRequest;

use Moo;
extends 'Streamtools::API::BaseRequest';
use namespace::clean;

1;


t/lib/Streamtools/API/TestRequest.pm
use strict;
use warnings;
package Streamtools::API::TestRequest;

use Moo;
extends 'Streamtools::API::BaseRequest';
use MooX::Types::MooseLike::Base qw(:all);
use namespace::clean;

has method => (
    is => 'rwp',
    isa => Enum['GET', 'POST', 'PUT', 'DELETE'],
    required => 0
);

has path => (
    is => 'rwp',
    isa => Str,
    required => 0
);

has headers => (
    is => 'rwp',
    isa => HashRef[Str],
    required => 0
);

has content => (
    is => 'rwp',
    isa => Str,
    required => 0
);

sub BUILD {
    my $self = shift;

    $self->_set_method('GET');
    $self->_set_path('frobnitzer?foo=baz');
    $self->_set_headers({ 'Content-type' => 'application/json' });
    $self->_set_content('');
}

1;


lib/Streamtools.pm
use strict;
use warnings;
package Streamtools;

use 5.018;


# ABSTRACT: Client library for Streamtools

=for test_synopsis
no strict 'vars'

=head1 SYNOPSIS

    use Streamtools;

    # Use the Streamtools server at http://localhost:7070
    my $st = Streamtools->new(server => 'localhost', port => 7070);


=head1 DESCRIPTION

This module provides a client library for the REST API provided by Streamtools.
From the developer blurb: "Streamtools is a graphical toolkit for dealing with
streams of data. Streamtools makes it easy to explore, analyse, modify and
learn from streams of data."

To find out more about Streamtools, please visit:

    http://nytlabs.github.io/streamtools/

Development of this code happens here:

    https://github.com/pfig/perl-streamtools/

=cut

use JSON 2.90;
use LWP::UserAgent::Determined 1.07;
use Readonly 2.00;
use Sub::Quote 1.006;

use Moo 1.006;
use MooX::Types::MooseLike::Base 0.27 qw(:all);
use namespace::clean 0.25;

has server => (
    is => 'ro',
    isa => Str,
    required => 1
);

has port => (
    is => 'ro',
    isa => Int,
    required => 1
);

has _ua => (
    is => 'lazy',
    isa => quote_sub(q{
        my ($val) = @_;
        die 'Not a LWP::UserAgent' unless (ref $val) =~ /^(Test::)?LWP::UserAgent/
    }),
    required => 0
);

has retry => (
    is => 'ro',
    isa => Bool,
    required => 0,
    default => 0
);

has timeout => (
    is => 'ro',
    isa => Int,
    required => 0,
    default => 10
);

has _json => (
    is => 'ro',
    isa => quote_sub(q{
        my ($val) = @_;
        die 'Not a JSON parser' unless (ref $val) =~ /^JSON/
    }),
    required => 0,
    default => sub { JSON->new->utf8; }
);

=head1 METHODS

=head2 new

Creates a new Streamtools object.

Takes two MANDATORY arguments, server and port:

    my $st = Streamtools->new(server => 'localhost', port => 7070);

Optionally you can specify whether to retry requests (retry => 1, defaults to
false), and what the connection timeout should be (timeout => 30, defaults to
10 seconds).

=head2 version

Get the version of a Streamtools service:

    my $version = $st->version;
    my $value = $version->request;

Alternatively, of course,

    my $version = $st->version->request;

=cut

sub version {
    my $self = shift;

    return Streamtools::API::Streamtools::Version->new(_st => $self);
}

=head2 library

Get the block library of a Streamtools service:

    my $library = $st->library;
    my $hashref = $library->request;

Or

    my $library = $st->library->request;

=cut

sub library {
    my $self = shift;

    return Streamtools::API::Streamtools::Library->new(_st => $self);
}

=head2 export

Get the current pattern in a Streamtools service:

    my $export = $st->export;
    my $hashref = $export->request;

Or

    my $export = $st->export->request;

=cut

sub export {
    my $self = shift;

    return Streamtools::API::Streamtools::Export->new(_st => $self);
}

# No user-serviceable parts below.

Readonly::Scalar my $KEEP_ALIVE_CACHESIZE => 10;

sub _build__ua {
    my $self = shift;

    my $ua;
    if ($self->retry) {
        $ua = LWP::UserAgent::Determined->new(
            keep_alive => $KEEP_ALIVE_CACHESIZE,
            requests_redirectable => [qw(GET HEAD DELETE PUT POST)],
        );
        $ua->timing('1,2,4,8,16,32');
    } else {
        $ua = LWP::UserAgent->new(
            keep_alive => $KEEP_ALIVE_CACHESIZE,
            requests_redirectable => [qw(GET HEAD DELETE PUT POST)],
        );
    }
    $ua->timeout($self->timeout);
    $ua->env_proxy;

    return $ua;
}

=head1 LICENSE

This library is free software and may be distributed under the same terms as
Perl itself. See L<http://dev.perl.org/licenses/>.


=head1 TESTING

Tests against a Streamtools server are only performed if the environment
variable LIVE_STREAMTOOLS_TESTS is set to the desired Streamtools instance's
URL (e.g., LIVE_STREAMTOOLS_TESTS=http://localhost:7070) - if it happens to
point to something other than a Streamtools server, these tests will fail.


=head1 AUTHOR

Pedro Figueiredo <pfig@me.com>


=head1 ACKNOWLEDGEMENTS

=over

=item * The Streamtools developers for a wonderful tool, in particular Jacqui
Maher and Eric Buth, who not only introduced me to Streamtools, but were kind
and patient enough to explain things and listen to my crazy ideas.

=back

=cut

1;


lib/Streamtools/API/BaseRequest.pm
use strict;
use warnings;
package Streamtools::API::BaseRequest;

# ABSTRACT: Base class for all API requests

=for test_synopsis
no strict 'vars'

=head1 SYNOPSIS

    package NewKidsOnTheBlock;
    extends 'Streamtools::API::BaseRequest';


=head1 DESCRIPTION

A base class for the different API requests.

Not for external use.

=cut

use HTTP::Headers 6.05;
use Sub::Quote 1.006;
use URI::Escape 1.65 qw(uri_escape_utf8);

use Moo 1.006;
with 'Streamtools::API::Requester';
use MooX::Types::MooseLike::Base 0.27 qw(:all);
use namespace::clean 0.25;

has _st => (
    is => 'ro',
    isa => quote_sub(q{
        my ($val) = @_;
        die 'Not a Streamtools object' unless (ref $val) =~ /^Streamtools/
    }),
    required => 1
);

has headers => (
    is => 'ro',
    isa => HashRef[Str],
    required => 0,
    default => sub { {}; }
);

has content => (
    is => 'ro',
    isa => Str,
    required => 0,
    default => sub { ''; }
);

sub method {
    die 'Please implement this method in your concrete classes';
}

sub path {
    die 'Please implement this method in your concrete classes';
}

sub request {
    die 'Please implement this method in your concrete classes';
}

#
# Make request objects
#
sub _make_request {
    my $self = shift;
    my $method = $self->method;
    my $path = $self->path;
    my $headers = $self->headers;
    my $content = $self->content;

    my $http_headers = $self->_make_headers($headers);
    my $protocol = 'http';
    my $uri = $protocol . '://' .
              $self->_st->server . ':' .
              $self->_st->port  . '/' .
              $self->_urlencode($path);

    my $request = HTTP::Request->new($method, $uri, $http_headers, $content);

    return $request;
}

#
# Make HTTP headers
#
sub _make_headers {
    my ($self, $headers) = @_;
    $headers ||= {};

    my $http_headers = HTTP::Headers->new;
    while (my ($k, $v) = each %$headers) {
        $http_headers->header($k => $v);
    }

    return $http_headers;
}

#
# URL-encode the final URL for the request
#
sub _urlencode {
    my ($self, $unencoded) = @_;
    return uri_escape_utf8($unencoded, '^A-Za-z0-9_-');
}

1;

=for test_synopsis
no strict 'vars'

=head1 SYNOPSIS

  my $request = Streamtools::API::BaseRequest->new(
  )->_make_request;

=head1 DESCRIPTION

This module is a base class for all the Streamtools API requests.

=head1 METHODS


lib/Streamtools/API/Streamtools/Export.pm
use strict;
use warnings;
package Streamtools::API::Streamtools::Export;

# ABSTRACT: GET /library

=for test_synopsis
no strict 'vars'

=head1 SYNOPSIS

    my $st = Streamtools->new(server => 'localhost', port => 7070);
    my $pattern = $st->export->request;

=head1 DESCRIPTION

A representation of the /export resource.

=cut

use Carp 1.3301;
use HTTP::Status 6.03 qw(:is status_message);
use Readonly 2.00;

use Moo 1.006;
use MooX::Types::MooseLike::Base 0.27 qw(:all);

extends 'Streamtools::API::BaseRequest';

use namespace::clean 0.25;

Readonly::Scalar my $METHOD => 'GET';
Readonly::Scalar my $PATH => 'export';

has method => (
    is => 'ro',
    isa => Str,
    required => 0,
    default => sub { $METHOD }
);

has path => (
    is => 'ro',
    isa => Str,
    required => 0,
    default => sub { $PATH }
);

sub request {
    my $self = shift;

    my $http_request = $self->_make_request;
    # TODO: proper error/exception handling
    my $response = $self->_st->_ua->request($http_request);
    my $code = $response->code;

    if (is_success($code)) {
        return $self->_st->_json->decode($response->decoded_content);
    }
    else {
        confess status_message($code);
    }
}

1;


lib/Streamtools/API/Streamtools/Version.pm
use strict;
use warnings;
package Streamtools::API::Streamtools::Version;

# ABSTRACT: GET /version

=for test_synopsis
no strict 'vars'

=head1 SYNOPSIS

    my $st = Streamtools->new(server => 'localhost', port => 7070);
    my $version = $st->version->request;

=head1 DESCRIPTION

A representation of the /version resource.

=cut

use Carp 1.3301;
use HTTP::Status 6.03 qw(:is status_message);
use Readonly 2.00;

use Moo 1.006;
use MooX::Types::MooseLike::Base 0.27 qw(:all);

extends 'Streamtools::API::BaseRequest';

use namespace::clean 0.25;

Readonly::Scalar my $METHOD => 'GET';
Readonly::Scalar my $PATH => 'version';

has method => (
    is => 'ro',
    isa => Str,
    required => 0,
    default => sub { $METHOD }
);

has path => (
    is => 'ro',
    isa => Str,
    required => 0,
    default => sub { $PATH }
);

sub request {
    my $self = shift;

    my $http_request = $self->_make_request;
    # TODO: proper error/exception handling
    my $response = $self->_st->_ua->request($http_request);
    my $code = $response->code;

    if (is_success($code)) {
        my $val = $self->_st->_json->decode($response->decoded_content);
        return $val->{'Version'};
    }
    else {
        confess status_message($code);
    }
}

1;


lib/Streamtools/API/Streamtools/Library.pm
use strict;
use warnings;
package Streamtools::API::Streamtools::Library;

# ABSTRACT: GET /library

=for test_synopsis
no strict 'vars'

=head1 SYNOPSIS

    my $st = Streamtools->new(server => 'localhost', port => 7070);
    my $library = $st->library->request;

=head1 DESCRIPTION

A representation of the /library resource.

=cut

use Carp 1.3301;
use HTTP::Status 6.03 qw(:is status_message);
use Readonly 2.00;

use Moo 1.006;
use MooX::Types::MooseLike::Base 0.27 qw(:all);

extends 'Streamtools::API::BaseRequest';

use namespace::clean 0.25;

Readonly::Scalar my $METHOD => 'GET';
Readonly::Scalar my $PATH => 'library';

has method => (
    is => 'ro',
    isa => Str,
    required => 0,
    default => sub { $METHOD }
);

has path => (
    is => 'ro',
    isa => Str,
    required => 0,
    default => sub { $PATH }
);

sub request {
    my $self = shift;

    my $http_request = $self->_make_request;
    # TODO: proper error/exception handling
    my $response = $self->_st->_ua->request($http_request);
    my $code = $response->code;

    if (is_success($code)) {
        return $self->_st->_json->decode($response->decoded_content);
    }
    else {
        confess status_message($code);
    }
}

1;
